### ISSUE TITLE
HNC: Reflect resource sync state within the SubnamespaceAnchor status

### ISSUE THREAD

**Issue by mgoltzsche** [kubernetes-sigs/multi-tenancy#1351]
Currently the only status indicator of a SubnamespaceAnchor is `status.state` which one has to wait for to become `Ok`. At that point the `Namespace` resource has been created but the `RoleBindings` haven't necessarily been synced from the parent into the child namespace yet and there is no status indicator for that a caller could use.
Thus, when programmatically creating a SubnamespaceAnchor and applying resources into the created child namespace afterwards, operations often fail when the corresponding RoleBinding that allows the caller to access the child namespace hasn't been synced into the child namespace yet.
Therefore I propose to either introduce another status field or condition that reflects the resource sync state or let `status.state` become `Ok` only after all resources (or at least the RoleBindings) have been synced into the child namespace.

**Comment by adrianludwin** [comment]
Thanks @mgoltzsche . The problem we'll run into is that the resource syncing is done by a bunch of asynchronous, independent threads (well, goroutines) and we don't actually get any kind of signal even internally as to when all of those operations are complete.

Also, if a resource in a parent namespace _changed_, would you want the `state` to switch to something like `Pending` before going back to `Ok` again?

Most importantly - are you seeing a long delay between when the subnamespace is created and when the resources have been copied? I would have expected that it would typically be quite fast (e.g. <1s). If that's the case, could you add a retry loop to your deployment process?

Thanks, A

**Comment by mgoltzsche** [comment]
To workaround this indeed I am going to poll for the RoleBinding first for a few seconds before I run the logic that depends on it.

However it would be great if `state` would become `Ok` only after all syncable resources that existed within the parent namespace at the time of creation have been syned. It does not need to switch to a `Pending` state afterwards when resources within the parent namespace change.

The RoleBinding was synced after ~5s last time I looked.

**Comment by adrianludwin** [comment]
Sorry for the delay in getting back to you, competing projects...

Instead of polling specifically for the RoleBinding, would it be possible to simply retry all errors? HNC itself takes this approach - it's based on https://github.com/kubernetes-sigs/controller-runtime which automatically implements a backoff and retry approach for all errors, which helps with a wide range of transient problems, including those caused by Kubernetes itself.

If we did want to implement this, I think we'd have to do something like the following:

- Move [`propagatedObjects`](https://github.com/kubernetes-sigs/multi-tenancy/blob/master/incubator/hnc/internal/reconcilers/object.go#L95) from a "global" map inside the object reconciler to a per-namespace map in the forest (e.g. [here](https://github.com/kubernetes-sigs/multi-tenancy/blob/master/incubator/hnc/internal/forest/namespaceobjects.go))
- Update the [`HNCConfiguration`](https://github.com/kubernetes-sigs/multi-tenancy/blob/d7a0f3e36c088739c32bd2db1435ad80c75fc733/incubator/hnc/internal/reconcilers/hnc_config.go#L406) to read the stats off the forest instead of from the object reconciler
- Whenever an object reconciler propagates a new object and adds it to the map, check to see if every object that _should_ be propagated into that namespace actually _is_ propagated into the namespace by looking through the forest (this can be done quickly because it's all in-memory). If it is, ping the hierarchy reconciler to update the statuses.

This is a non-trivial bit of work but seems like it could be plausible. My biggest concern here would be having things get out of sync and never flipping the status from `not ready` to `ready` (or vice versa - this change could easily give us the inverse).

Would you be interested in developing a patch? I think for a change of this size we'd want to have a small design doc and test plan to make sure the feature actually works.

Thanks!

**Comment by mgoltzsche** [comment]
My concrete problem was fixed by polling the RoleBinding. In my case retrying all errors is not really an option since the logic that runs afterwards is pretty complex, should fail early on errors and has different timeout requirements which I don't want to mix up with the timeout I am applying to wait for the RoleBinding.
Also I wanted to avoid polling the RoleBinding because it made me maintain an additional configuration field for its name which I thought I wouldn't need otherwise. However - at least in my case - it turned out that the configuration of the RoleBinding name serves another purpose: After it has been synced I copy the RoleBinding to bind it to a ServiceAccount within the child namespace in order to deploy the manifests using the child namespace's ServiceAccount token to make sure the manifests/access cannot leak into another namespace. It seems to me that this is a common pattern to apply generic manifests into an instantly created child namespace safely - and this should be implemented using the retry-on-error approach.

Therefore I don't need the feature I requested here anymore and will close the issue. In case somebody else needs this it can be reopened.
Thanks for explaining how it could be implemented within HNC anyway!
